cw/cw305/waverunner.py

```diff
@@ -0,0 +1,246 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+"""Support for capturing traces using LeCroy WaveRunner 9104."""
+
+import logging
+import re
+
+import numpy as np
+import vxi11
+
+
+class _Timeout:
+    """Helper class for setting scoped timeout values."""
+
+    def __init__(self, instr, timeout):
+        self._instr = instr
+        self._orig_timeout = self._instr.timeout
+        self._instr.timeout = timeout
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self._instr.timeout = self._orig_timeout
+
+
+class WaveRunner:
+    """Class for capturing traces using a WaveRunner oscilloscope.
+
+    This class operates the oscilloscope in sequence mode to improve performance and in
+    2-channel mode to maximize sampling rate. Trigger and power signals must be
+    connected to channels 2 and 3, respectively.
+
+    This class is only tested to work with WaveRunner 9104 series.
+
+    For more information on the commands used in this module please see:
+    - Operator's Manual WaveRunner 9000 and WaveRunner 8000-R Oscilloscopes
+        (http://cdn.teledynelecroy.com/files/manuals/waverunner-9000-operators-manual.pdf)
+    - MAUI Oscilloscopes Remote Control and Automation Manual
+        (http://cdn.teledynelecroy.com/files/manuals/maui-remote-control-and-automation-manual.pdf)
+
+    Typical usage:
+    >>> waverunner = WaveRunner('192.168.1.227')
+    >>> waverunner.configure()
+    >>> while foo:
+    >>>     ...
+    >>>     waverunner.seq_num_waves = seq_num_waves
+    >>>     waverunner.arm()
+    >>>     waves = waverunner.wait_for_acquisition_and_transfer_waves()
+    >>>     ...
+
+    Attributes:
+        seq_num_waves: Number of waves per sequence.
```

Comment:
nit: Can you elaborate a bit more about this? Is each wave equivalent to num_samples, and the scope just does seq_num_waves * num_samples after the trigger?

Reply:
Each wave has num_samples sample, we send a single command to the oscilloscope to put in sequence mode (instead of talking to it for each wave) and it captures a total of seq_num_waves waves, each starting with a trigger. I've seen this being called segmented traces, repeated trigger, etc.

Reply:
Updated class docstring.